docs/ARCHITECTURE.md

```diff
@@ -0,0 +1,62 @@
+The purpose of this file is to describe the functional and key architecture of
+this project. It is spiced with some rational throughout.
+
+# Slots & Tokens
+
+This library shall provide multiple tokens for the user(s). Any given slot can
+only contain one token at any given time. In case of smartcards the slot is a
+reader whilst the token is the actual smartcard. Since this project won't
+implement a separate tool for "virtual swapping" of tokens, each token is
+associated with its own slot. Thus, the library offers a set of slots with a
+token inside each of these.
+
+In order to allow the creation of new tokens using the pkcs11 means, this
+library presents one more slot/token pair than tokens currently exist. The
+additional token is a blank token that can be initialized using the C_InitToken
+function.
+(Note: This is not implemented yet)
+
+# Objects
+
+Each token can contain an arbitrary amount of keys and other object. (Note:
+other objects are not implemented yet). The keys are TPM-based keys, i.e. they
+are only known to the TPM and also the key operations are executed inside the
+TPM. The other objects are encrypted using hybrid encryption with the outer
+key being sealed to the TPM.
+
+Note: Since the TPM's object are stored (encrypted) external to the TPM on the
+HDD, it is not possible to delete them, if your attacker model allows access to
+the HDD.
+
+# Key Authorizations
+
+For each token, two TPM seal blobs exist one for the SO and one for the User.
+The authValue for these seal blobs are SO's PIN and the User's PIN. The data
```

Comment:
yes, but:
  - respective to the so and user seal objects.
  - the pins are fed through a pbkdf and the output of that is the seal object authvalue.

Reply:
Does using PBKDF add anything here? I'm not strictly opposed to using it, but as far as I can tell, it only offers two features:

1. It's slow, which helps to slow down brute force; but that should be handled anyway by the TPM's DA protection (or using PIN Fail NV Indices).

2. It compresses long PINs so that they're smaller than the key's NameAlg hash output size (as auth values need to be). But this could be handled by a single hash iteration, and could be omitted for inputs that fit.

Reply:
Really just 1 because as you point out 2. It made a little more sense when the db and files were separate because it made the attacker need both the file and db, but now it just requires the db. So in the off chance someone did something sloppy it threw a chair in the hallway.